### PR TITLE
Lightstep gPRC generation for tracing

### PR DESCRIPTION
**Pull request by RomanDzhabarov**
Make LightStepSink to send gPRC requests to lightstep collectors.

**not for this iteration** 2) Right now we make gRPC call only when buffer is full, for the first iteration I'd like to keep it this way, but we might need flush thread to flush traces for low rps services

### CODE REVIEW

File: source/common/CMakeLists.txt [hunk 1]
```diff
@@ -47,6 +47,7 @@ add_library(
   grpc/common.cc
   grpc/http1_bridge_filter.cc
   grpc/rpc_channel_impl.cc
+  grpc/utility.cc
```
Comment (mattklein123) [inline source/common/CMakeLists.txt]: nit: I think just add this code to common.cc, probably don't need a new source file

File: source/common/common/utility.h [hunk 2]
```diff
@@ -92,4 +92,10 @@ class StringUtil {
    */
   static bool startsWith(const std::string& source, const std::string& start,
                          bool case_sensitive = true);
+
+  /**
+   * @return original @param input string if it's not empty or @param default_value otherwise.
+   */
+  static const std::string& valueOrDefault(const std::string& input,
+                                           const std::string& default_value = "-");
```
Comment (mattklein123) [inline source/common/common/utility.h]: nit: having a default param of "-" for a common utility function is a little strange, I would just have no default param.

File: source/common/tracing/http_tracer_impl.cc [hunk 3]
```diff
@@ -113,46 +114,56 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
...
+      service_node_(service_node), options_(options) {
+  if (!cm_.get(collector_cluster_)) {
+    throw EnvoyException(fmt::format("{} collector cluster is not defined on cluster manager level",
+                                     collector_cluster_));
   }
+}
 
-  return join_ids;
+lightstep::Tracer& LightStepSink::thread_local_tracer() {
+  static thread_local lightstep::Tracer tracer(lightstep::NewUserDefinedTransportLightStepTracer(
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: need to use envoy tls here for testing (probably obvious but pointing out anyway)

**Comment by mattklein123** [comment]
@RomanDzhabarov there are some more nits and a bunch of things to clean up (which you probably already know about) but in general looks like the right direction. I will wait until you have it more complete with tests then will look again.

File: source/common/CMakeLists.txt (hunk 1 above)
Comment (RomanDzhabarov) [inline source/common/CMakeLists.txt]: sounds good, moving that to common.

File: source/common/common/utility.h (hunk 2 above)
Comment (RomanDzhabarov) [inline source/common/common/utility.h]: fixed.

File: source/common/tracing/http_tracer_impl.cc [hunk 4]
```diff
@@ -113,46 +114,56 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
...
+void LightStepRecorder::RecordSpan(lightstep::collector::Span&& span) {
+  builder_.addSpan(std::move(span));
 
-  return msg;
+  const int min_span_number_to_send = 5;
+  if (builder_.pendingSpans() == min_span_number_to_send) {
+    lightstep::collector::ReportRequest request;
+    std::swap(request, builder_.pending());
+
+    Grpc::RpcAsyncClientImpl client(sink_->clusterManager());
```
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: fixed, moved logic to sink

File: source/common/tracing/http_tracer_impl.cc (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: fixed

File: include/envoy/grpc/rpc_channel.h [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 #include "envoy/common/optional.h"
 #include "envoy/common/pure.h"
+#include "envoy/http/async_client.h"
```
Comment (mattklein123) [inline include/envoy/grpc/rpc_channel.h]: delete

Comment (RomanDzhabarov) [inline include/envoy/grpc/rpc_channel.h]: removed

File: source/common/tracing/http_tracer_impl.cc [hunk 6]
```diff
@@ -113,46 +111,68 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
   }
 }
 
-Http::MessagePtr LightStepUtility::buildHeaders(const std::string& access_token) {
-  Http::MessagePtr msg{new Http::RequestMessageImpl()};
+LightStepRecorder::LightStepRecorder(const lightstep::TracerImpl& tracer, LightStepSink* sink)
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: nit: LightStepSink& sink if possible (might not be due to bind call)

**Comment by mattklein123** [comment]
this looks good, but let's try to actually test that we can send spans before we commit

File: source/common/tracing/http_tracer_impl.cc (hunk 6 above)
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: fixed

**Comment by RomanDzhabarov** [comment]
Tested on actual collector, works fine. There were issues on lightstep side, but they fixed it temporary on server side. They will ship new version of client library, so that it'll be a long term fix.
Let merge these changes, and upgrade to new lightstep client library once they have it.

File: source/common/grpc/common.h [hunk 7]
```diff
@@ -33,9 +43,18 @@ class Common {
                                          const std::string& service_full_name,
                                          const std::string& method_name);
 
+  /**
+   * Basic validation of gRPC response, @throws Grpc::Exception in case of non successful response.
+   *
```
Comment (mattklein123) [inline source/common/grpc/common.h]: nit: extra line

File: source/common/tracing/http_tracer_impl.cc [hunk 8]
```diff
@@ -218,9 +222,23 @@ void LightStepSink::flushTrace(const Http::HeaderMap& request_headers, const Htt
...
-void LightStepRecorder::onSuccess(Http::MessagePtr&&) { sink_.stats().collector_success_.inc(); }
+void LightStepRecorder::onSuccess(Http::MessagePtr&& msg) {
+  try {
+    Grpc::Common::validateResponse(*msg);
+
+    Grpc::Common::chargeStat(sink_.statsStore(), sink_.collectorCluster(),
+                             LightStepSink::LIGHTSTEP_SERVICE, LightStepSink::LIGHTSTEP_METHOD,
+                             true);
+  } catch (const Grpc::Exception& ex) {
+    std::cout << "ex: " << ex.what() << std::endl;
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: delete

File: source/common/tracing/http_tracer_impl.cc [hunk 9]
```diff
@@ -218,9 +222,23 @@ void LightStepSink::flushTrace(const Http::HeaderMap& request_headers, const Htt
...
+    Grpc::Common::chargeStat(sink_.statsStore(), sink_.collectorCluster(),
+                             "lightstep.collector.CollectorService", "Report", false);
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: constants

File: test/common/tracing/http_tracer_impl_test.cc [hunk 10]
```diff
@@ -272,16 +272,13 @@ TEST(HttpNullTracerTest, NoFailures) {
...
-      : stats_{LIGHTSTEP_STATS(POOL_COUNTER_PREFIX(fake_stats_, "prefix.tracing.lightstep."))} {}
-
   void setup(Json::Object& config) {
     std::unique_ptr<lightstep::TracerOptions> opts(new lightstep::TracerOptions());
     opts->access_token = "sample_token";
     opts->tracer_attributes["lightstep.guid"] = "random_guid";
     opts->tracer_attributes["lightstep.component_name"] = "component";
 
-    sink_.reset(new LightStepSink(config, cm_, "prefix.", fake_stats_, "service_node", tls_,
+    sink_.reset(new LightStepSink(config, cm_, "not_used_now", stats_, "service_node", tls_,
```
Comment (mattklein123) [inline test/common/tracing/http_tracer_impl_test.cc]: just delete the parameter ("not_used_now")

File: source/common/grpc/common.h (hunk 7 above)
Comment (RomanDzhabarov) [inline source/common/grpc/common.h]: done

File: source/common/tracing/http_tracer_impl.cc (hunk 8 above)
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: done

File: source/common/tracing/http_tracer_impl.cc (hunk 9 above)
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: oops

File: test/common/tracing/http_tracer_impl_test.cc (hunk 10 above)
Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: i though about using it later, but will remove for now.

Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: done

**Pull request merged**